builtin/logical/totp/path_keys.go

```diff
@@ -0,0 +1,400 @@
+	accountName := data.Get("account_name").(string)
+	period := data.Get("period").(int)
+	algorithm := data.Get("algorithm").(string)
+	digits := data.Get("digits").(int)
+	skew := data.Get("skew").(int)
+	qrSize := data.Get("qr_size").(int)
+	keySize := data.Get("key_size").(int)
+	inputURL := data.Get("url").(string)
+
+	if generate == true {
+		if keyString != "" {
+			return logical.ErrorResponse("a key should not be passed if generate is true"), nil
+		}
+		if inputURL != "" {
+			return logical.ErrorResponse("a url should not be passed if generate is true"), nil
+		}
+	}
+
+	// Read parameters from url if given
+	if inputURL != "" {
+		//Parse url
+		urlObject, err := url.Parse(inputURL)
+		if err != nil {
+			return logical.ErrorResponse("an error occured while parsing url string"), nil
+		}
+
+		//Set up query object
+		urlQuery := urlObject.Query()
+		path := strings.TrimPrefix(urlObject.Path, "/")
+		index := strings.Index(path, ":")
+
+		//Read issuer
+		urlIssuer := urlQuery.Get("issuer")
+		if urlIssuer != "" {
+			issuer = urlIssuer
+		} else {
+			if index != -1 {
+				issuer = path[:index]
+			}
+		}
+
+		//Read account name
+		if index == -1 {
+			accountName = path
+		} else {
+			accountName = path[index+1:]
+		}
+
+		//Read key string
+		keyString = urlQuery.Get("secret")
+
+		//Read period
+		periodQuery, err := strconv.Atoi(urlQuery.Get("period"))
+		if err == nil {
+			period = periodQuery
+		}
+
+		//Read digits
+		digitsQuery, err := strconv.Atoi(urlQuery.Get("digits"))
+		if err == nil {
+			digits = digitsQuery
+		}
+
+		//Read algorithm
+		algorithmQuery := urlQuery.Get("algorithm")
+		if algorithmQuery != "" {
+			algorithm = algorithmQuery
+		}
+	}
+
+	// Translate digits and algorithm to a format the totp library understands
+	var keyDigits otplib.Digits
+	switch digits {
+	case 6:
+		keyDigits = otplib.DigitsSix
+	case 8:
+		keyDigits = otplib.DigitsEight
+	default:
+		return logical.ErrorResponse("the digit value can only be 6 or 8"), nil
+	}
+
+	var keyAlgorithm otplib.Algorithm
+	switch algorithm {
+	case "SHA1":
+		keyAlgorithm = otplib.AlgorithmSHA1
+	case "SHA256":
+		keyAlgorithm = otplib.AlgorithmSHA256
+	case "SHA512":
+		keyAlgorithm = otplib.AlgorithmSHA512
+	default:
+		return logical.ErrorResponse("the algorithm value is not valid"), nil
+	}
+
+	// Enforce input value requirements
+	if period <= 0 {
+		return logical.ErrorResponse("the period value must be greater than zero"), nil
+	}
+
+	switch skew {
+	case 0:
+	case 1:
+	default:
+		return logical.ErrorResponse("the skew value must be 0 or 1"), nil
+	}
+
+	if qrSize <= 0 {
+		return logical.ErrorResponse("the qr_size value must be greater than zero"), nil
+	}
+
+	if keySize <= 0 {
+		return logical.ErrorResponse("the key_size value must be greater than zero"), nil
+	}
+
+	// Period, Skew and Key Size need to be unsigned ints
+	uintPeriod := uint(period)
+	uintSkew := uint(skew)
+	uintKeySize := uint(keySize)
+
+	var response logical.Response
```

Comment:
Make this a pointer -- the default value will be nil. That way at the bottom you don't need an `if` block to switch on `generate` -- you just return `response` and if `generate` was false it'll automatically be `nil`.